### PR TITLE
[explorer] - add epochs pages

### PR DESCRIPTION
**Pull request by mamos-mysten**
## Description 

Adds Epochs pages to Explorer 

## Test Plan 

---
If your changes are not user-facing and not a breaking change, you can skip the following section. Otherwise, please indicate what changed, and then add to the Release Notes section as highlighted during the release process.

### Type of Change (Check all that apply)

- [ ] user-visible impact
- [ ] breaking change for a client SDKs
- [ ] breaking change for FNs (FN binary must upgrade)
- [ ] breaking change for validators or node operators (must upgrade binaries)
- [ ] breaking change for on-chain data layout
- [ ] necessitate either a data wipe or data migration

### CODE REVIEW

**Comment by mamos-mysten** [comment]
Current dependencies on/for this PR:
* main
  * **PR #10033** <a href="https://app.graphite.dev/github/pr/MystenLabs/sui/10033" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/MystenLabs/sui/10033?utm_source=stack-comment).

File: apps/explorer/src/components/Activity/index.tsx [hunk 1]
```diff
@@ -20,13 +21,18 @@ const AUTO_REFRESH_ID = 'auto-refresh';
 const REFETCH_INTERVAL_SECONDS = 10;
 const REFETCH_INTERVAL = REFETCH_INTERVAL_SECONDS * 1000;
 
+const tabs: Record<string, number> = {
```
Comment (Jordan-Mysten) [inline apps/explorer/src/components/Activity/index.tsx:24]: follow-up: this is basically a worse array

File: sdk/typescript/src/types/epochs.ts [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+export const EpochInfo = object({
+  epoch: number(),
+  validators: array(SuiValidatorSummary),
+  epochTotalTransactions: number(),
+  firstCheckpointId: number(),
+  epochStartTimestamp: number(),
+  endOfEpochInfo: nullable(EndOfEpochInfo),
+});
+export type EpochInfo = Infer<typeof EpochInfo>;
+export const EpochPage = object({
```
Comment (Jordan-Mysten) [inline sdk/typescript/src/types/epochs.ts:39]: man really said no newlines

File: sdk/typescript/src/providers/json-rpc-provider.ts [hunk 3]
```diff
@@ -821,4 +822,30 @@ export class JsonRpcProvider {
       NetworkMetrics,
     );
   }
+  /**
+   * Return the committee information for the asked epoch
+   */
+  async getEpochs(input?: {
+    cursor?: number;
```
Comment (Jordan-Mysten) [inline sdk/typescript/src/providers/json-rpc-provider.ts:829]: follow-up: I think you can use the PaginationOptions thing?

File: apps/explorer/src/ui/RingChart.tsx [hunk 4]
```diff
@@ -26,7 +26,7 @@ function Legend({ data, title }: Pick<RingChartProps, 'data' | 'title'>) {
                     <div className="flex items-center gap-1.5" key={label}>
                         <div
                             style={{ backgroundColor: color }}
-                            className="h-3 w-3 rounded-full"
+                            className="h-3 w-3 rounded-sm"
```
Comment (Jordan-Mysten) [inline apps/explorer/src/ui/RingChart.tsx:29]: curious why

File: apps/explorer/src/ui/Pagination.tsx [hunk 5]
```diff
@@ -21,6 +21,44 @@ export interface PaginationResponse<Cursor = string> {
     hasNextPage: boolean;
 }
 
+// todo: temporary solution to allow pagination for queries
```
Comment (Jordan-Mysten) [inline apps/explorer/src/ui/Pagination.tsx:24]: temporary until what? is this actually going to change?

File: apps/explorer/src/ui/InternalLink.tsx [hunk 6]
```diff
@@ -30,11 +30,12 @@ function createInternalLink<T extends string>(
     };
 }
 
+export const EpochLink = createInternalLink('epoch', 'epoch', (epoch) => epoch);
 export const CheckpointLink = createInternalLink('checkpoint', 'digest');
 export const CheckpointSequenceLink = createInternalLink(
     'checkpoint',
     'sequence',
-    (address: string) => address
+    (sequence: string) => (sequence ? `#${sequence}` : '')
```
Comment (Jordan-Mysten) [inline apps/explorer/src/ui/InternalLink.tsx:38]: just to confirm: we want this treatment everywhere

File: apps/explorer/src/pages/checkpoints/CheckpointsTable.tsx [hunk 7]
```diff
@@ -2,52 +2,73 @@
...
 export function CheckpointsTable({
     initialLimit,
+    initialCursor,
+    maxCursor,
     disablePagination,
     refetchInterval,
 }: CheckpointsTableProps) {
     const rpc = useRpcClient();
     const [limit, setLimit] = useState(initialLimit);
+    const [cursor, setCursor] = useState(initialCursor);
```
Comment (Jordan-Mysten) [inline apps/explorer/src/pages/checkpoints/CheckpointsTable.tsx:33]: This is now tracked in two places??

File: apps/explorer/src/pages/checkpoints/CheckpointsTable.tsx [hunk 8]
```diff
@@ -2,52 +2,73 @@
...
+    }, [countQuery.data, initialCursor, maxCursor]);
 
     const { data: checkpointsData } = useQuery(
-        ['checkpoints', { limit, cursor: pagination.cursor }],
+        ['checkpoints', { limit, cursor }],
         () =>
             rpc.getCheckpoints({
-                limit,
+                limit:
+                    cursor && maxCursor && +cursor - +limit < +maxCursor
```
Comment (Jordan-Mysten) [inline apps/explorer/src/pages/checkpoints/CheckpointsTable.tsx:53]: No idea what this is. Lots of type conversion here too. Almost certainly needs a comment or two

File: apps/explorer/src/pages/checkpoints/CheckpointsTable.tsx [hunk 9]
```diff
@@ -2,52 +2,73 @@
...
             }),
         {
             keepPreviousData: true,
             // Disable refetching if not on the first page:
-            refetchInterval: pagination.cursor ? undefined : refetchInterval,
+            refetchInterval: cursor ? undefined : refetchInterval,
         }
     );
 
+    useEffect(() => {
```
Comment (Jordan-Mysten) [inline apps/explorer/src/pages/checkpoints/CheckpointsTable.tsx:66]: follow-up: this is bad, doesn't seem required at all

**Review by Jordan-Mysten — APPROVED** [review]

**Pull request merged**